### DISCUSSION TITLE
TB can't save best model.

### DISCUSSION THREAD

**Discussion by Raizenr** [coqui-ai/TTS#494] [General Q&A]
**Nothing in /app/data/TTS/tbloger/:**
  
      root@e82bbb799dd4:**/app/data/TTS/tbloger# ls -la**
      total 8
      drwxr-xr-x  2 root root 4096 May 20 07:59 .
      drwxr-xr-x 11 root root 4096 May 20 06:36 ..
      root@e82bbb799dd4:/app/data/TTS/tbloger#

**then i have to start train vocoder:**  

    python /app/data/TTS/TTS/bin/distribute.py --script train_vocoder_wavernn.py --config_path /app/data/wavernn_config.json
    root@e82bbb799dd4:/app/data/TTS/tbloger# ls -la
    total 12
    drwxr-xr-x  3 root root 4096 May 20 08:15 .
    drwxr-xr-x 11 root root 4096 May 20 06:36 ..
    drwxrwxr-x  3 root root 4096 May 20 08:15 wavernn_librittts-May-20-2021_08+15AM-d03aa36
    root@e82bbb799dd4:/app/data/TTS/tbloger#

**It's already created tb folder.**   
 
    root@e82bbb799dd4:/app/data/TTS/tbloger# cd wavernn_librittts-May-20-2021_08+15AM-d03aa36/
    root@e82bbb799dd4:/app/data/TTS/tbloger/wavernn_librittts-May-20-2021_08+15AM-d03aa36# ls -la
    total 32
    drwxrwxr-x 3 root root  4096 May 20 08:15 .
    drwxr-xr-x 3 root root  4096 May 20 08:15 ..
    -rw-r--r-- 1 root root  5422 May 20 08:15 config.json
    -rw-r--r-- 1 root root     0 May 20 08:15 events.out.tfevents.1621498539.e82bbb799dd4
    -rw-r--r-- 1 root root 10432 May 20 08:15 scale_stats.npy
    drwxrwxr-x 2 root root  4096 May 20 08:15 test_audios

**Then it try to save best model in tb folder it's failed and remove created directory**

      > TRAINING (2021-05-20 08:19:12)
    Traceback (most recent call last):
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 445, in <module>
        main(args)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 417, in main
        _, global_step = train(model_wavernn, optimizer, criterion, scheduler, scaler, ap, global_step, epoch)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 166, in train
        tb_logger.tb_train_iter_stats(global_step, iter_stats)
    AttributeError: 'NoneType' object has no attribute 'tb_train_iter_stats'
    Traceback (most recent call last):
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 445, in <module>
        main(args)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 417, in main
        _, global_step = train(model_wavernn, optimizer, criterion, scheduler, scaler, ap, global_step, epoch)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 166, in train
        tb_logger.tb_train_iter_stats(global_step, iter_stats)
    AttributeError: 'NoneType' object has no attribute 'tb_train_iter_stats'
    Traceback (most recent call last):
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 445, in <module>
        main(args)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 417, in main
        _, global_step = train(model_wavernn, optimizer, criterion, scheduler, scaler, ap, global_step, epoch)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 166, in train
        tb_logger.tb_train_iter_stats(global_step, iter_stats)
    AttributeError: 'NoneType' object has no attribute 'tb_train_iter_stats'
    Traceback (most recent call last):
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 445, in <module>
        main(args)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 417, in main
        _, global_step = train(model_wavernn, optimizer, criterion, scheduler, scaler, ap, global_step, epoch)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 166, in train
        tb_logger.tb_train_iter_stats(global_step, iter_stats)
    AttributeError: 'NoneType' object has no attribute 'tb_train_iter_stats'
    Traceback (most recent call last):
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 445, in <module>
        main(args)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 417, in main
        _, global_step = train(model_wavernn, optimizer, criterion, scheduler, scaler, ap, global_step, epoch)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 166, in train
        tb_logger.tb_train_iter_stats(global_step, iter_stats)
    AttributeError: 'NoneType' object has no attribute 'tb_train_iter_stats'
    Traceback (most recent call last):
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 445, in <module>
        main(args)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 417, in main
        _, global_step = train(model_wavernn, optimizer, criterion, scheduler, scaler, ap, global_step, epoch)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 166, in train
        tb_logger.tb_train_iter_stats(global_step, iter_stats)
    AttributeError: 'NoneType' object has no attribute 'tb_train_iter_stats'
    Traceback (most recent call last):
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 445, in <module>
        main(args)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 417, in main
        _, global_step = train(model_wavernn, optimizer, criterion, scheduler, scaler, ap, global_step, epoch)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 166, in train
        tb_logger.tb_train_iter_stats(global_step, iter_stats)
    AttributeError: 'NoneType' object has no attribute 'tb_train_iter_stats'
    
       --> STEP: 24/42 -- GLOBAL_STEP: 25
         | > model_loss: 11.70126  (11.74162)
         | > step_time: 0.24
         | > loader_time: 0.0007
         | > current_lr: 0.0001
    
       --> STEP: 49/42 -- GLOBAL_STEP: 50
         | > model_loss: 11.46611  (11.66872)
         | > step_time: 0.21
         | > loader_time: 0.0005
         | > current_lr: 0.0001
    
       --> STEP: 74/42 -- GLOBAL_STEP: 75
         | > model_loss: 10.86477  (11.49251)
         | > step_time: 0.23
         | > loader_time: 0.0007
         | > current_lr: 0.0001
    
       --> STEP: 99/42 -- GLOBAL_STEP: 100
         | > model_loss: 10.45827  (11.26338)
         | > step_time: 0.21
         | > loader_time: 0.0007
         | > current_lr: 0.0001
    
       --> STEP: 124/42 -- GLOBAL_STEP: 125
         | > model_loss: 10.42513  (11.07053)
         | > step_time: 0.23
         | > loader_time: 0.0005
         | > current_lr: 0.0001
    
       --> STEP: 149/42 -- GLOBAL_STEP: 150
         | > model_loss: 10.07767  (10.93089)
         | > step_time: 0.27
         | > loader_time: 0.0007
         | > current_lr: 0.0001
    
       --> STEP: 174/42 -- GLOBAL_STEP: 175
         | > model_loss: 9.70348  (10.80107)
         | > step_time: 0.18
         | > loader_time: 0.0007
         | > current_lr: 0.0001
    
       --> STEP: 199/42 -- GLOBAL_STEP: 200
         | > model_loss: 10.04591  (10.69060)
         | > step_time: 0.18
         | > loader_time: 0.0008
         | > current_lr: 0.0001
    
       --> STEP: 224/42 -- GLOBAL_STEP: 225
         | > model_loss: 9.85868  (10.59747)
         | > step_time: 0.31
         | > loader_time: 0.0007
         | > current_lr: 0.0001
    
       --> STEP: 249/42 -- GLOBAL_STEP: 250
         | > model_loss: 9.72437  (10.51810)
         | > step_time: 0.27
         | > loader_time: 0.0009
         | > current_lr: 0.0001
    
       --> STEP: 274/42 -- GLOBAL_STEP: 275
         | > model_loss: 9.69591  (10.44518)
         | > step_time: 0.28
         | > loader_time: 0.0004
         | > current_lr: 0.0001
    
       --> STEP: 299/42 -- GLOBAL_STEP: 300
         | > model_loss: 9.66393  (10.39078)
         | > step_time: 0.34
         | > loader_time: 0.0007
         | > current_lr: 0.0001
    
       --> STEP: 324/42 -- GLOBAL_STEP: 325
         | > model_loss: 9.39969  (10.33003)
         | > step_time: 0.28
         | > loader_time: 0.0007
         | > current_lr: 0.0001
    
       --> TRAIN PERFORMACE -- EPOCH TIME: 93.65 sec -- GLOBAL_STEP: 338
         | > avg_model_loss: 10.30307
         | > avg_loader_time: 0.00119
         | > avg_step_time: 0.27219
    
     > EVALUATION
    
      --> EVAL PERFORMANCE
         | > avg_model_loss: 9.71504 (+0.00000)
         | > avg_loader_time: 1.21791 (+0.00000)
         | > avg_step_time: 0.09125 (+0.00000)
    
     > **BEST MODEL : /app/data/TTS/tbloger/wavernn_librittts-May-20-2021_08+19AM-d03aa36/best_model_338.pth.tar
    Traceback (most recent call last):**
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 445, in <module>
        main(args)
      File "/app/data/TTS/TTS/bin/train_vocoder_wavernn.py", line 436, in main
        scaler=scaler.state_dict() if c.mixed_precision else None,
      File "/app/data/TTS/TTS/vocoder/utils/io.py", line 112, in save_best_model
        **kwargs,
      File "/app/data/TTS/TTS/vocoder/utils/io.py", line 49, in save_model
        torch.save(state, output_path)
      File "/usr/local/lib/python3.6/dist-packages/torch/serialization.py", line 369, in save
        with _open_file_like(f, 'wb') as opened_file:
      File "/usr/local/lib/python3.6/dist-packages/torch/serialization.py", line 230, in _open_file_like
        return _open_file(name_or_buffer, mode)
      File "/usr/local/lib/python3.6/dist-packages/torch/serialization.py", line 211, in __init__
        super(_open_file, self).__init__(open(name, mode))
    **FileNotFoundError: [Errno 2] No such file or directory: '/app/data/TTS/tbloger/wavernn_librittts-May-20-2021_08+19AM-d03aa36/best_model_338.pth.tar'**

When i trained tacotron and glow-tts everything was fine.

Why do it happened? I guess it some problem with saving best model. And i don't understand why this folder delete. Can't found code in repository.

**Comment by erogol** [comment]
What version of TTS on what system?  Are you sure the path is in reach?

**Reply by Raizenr** [reply]
> What version of TTS on what system? Are you sure the path is in reach?

It's TTS 0.13 inside Docker Ubuntu 16.4. I has changed path to another 2 more times and check it. Folder is correct.

**Reply by erogol** [reply]
To help there should be something I can reproduce otherwise it is not something I see the cause for it. All these training scripts already tested on CI and they pass the tests.

**Comment by lexkoro** [comment]
```
tb_logger.tb_train_iter_stats(global_step, iter_stats)
AttributeError: 'NoneType' object has no attribute 'tb_train_iter_stats'
```

Any changes you've made to the script? Seems like you are missing an initialization of tb_logger.
